src/gsi/amassaeromod.f90

```diff
@@ -0,0 +1,191 @@
+   call gsi_bundlegetpointer (sval,trim(aerosols(ic)),sv_rank3,istatus)
+   if (istatus/=0) cycle
+   sv_rank3=zero
+   if(imodes_cmaq_fv3(ic).eq.1)then
+   do k=1,nsig
+      do j=1,lon2
+         do i=1,lat2
+            sv_rank3(i,j,k)=cv_amassi(i,j,k)*ges_aero(i,j,k)/ges_amassi(i,j,k)
+         end do
+      end do
+   end do
+   elseif(imodes_cmaq_fv3(ic).eq.2)then
+   do k=1,nsig
+      do j=1,lon2
+         do i=1,lat2
+            sv_rank3(i,j,k)=cv_amassj(i,j,k)*ges_aero(i,j,k)/ges_amassj(i,j,k)
+         if(ges_aero(i,j,k)/ges_amassj(i,j,k).ge.1.0)then
+           !print*,"amas2aero_tl:wg.ge.1 ",trim(aerosols(ic)),ges_aero(i,j,k),ges_amassj(i,j,k)
+         end if
+         end do
+      end do
+   end do
+   else
+   do k=1,nsig
+      do j=1,lon2
+         do i=1,lat2
+            sv_rank3(i,j,k)=cv_amassk(i,j,k)*ges_aero(i,j,k)/ges_amassk(i,j,k)
+         end do
+      end do
+   end do
+
+   end if
+   !print*,"amas2aero_tl: ",trim(aerosols(ic)),sv_rank3(1,1,1)
+end do
+
+return
+end subroutine amass2aero_tl
+
+subroutine amass2aero_ad(rval,wbundle,aerosols,naerosols)
+implicit none
+
+! Declare passed variables
+type(gsi_bundle),intent(in):: rval
+type(gsi_bundle),intent(inout):: wbundle
+integer(i_kind),intent(in) :: naerosols
+character(len=max_varname_length),intent(in):: aerosols(naerosols)
+
+! Declare local variables
+integer(i_kind) i,j,k,ic,istatus,it,ier
+real(r_kind),dimension(lat2,lon2,nsig) :: work
+real(r_kind),pointer,dimension(:,:,:) :: sv_rank3
+
+real(r_kind),pointer,dimension(:,:,:) :: cv_amassi,cv_amassj,cv_amassk
+
+    real(r_kind),dimension(:,:,:),pointer::ges_aero=>NULL()
+
+    real(r_kind),dimension(:,:,:),pointer::ges_amassi=>NULL()
+    real(r_kind),dimension(:,:,:),pointer::ges_amassj=>NULL()
+    real(r_kind),dimension(:,:,:),pointer::ges_amassk=>NULL()
+
+    ier = 0.0
+    it= ntguessig
+    call GSI_BundleGetPointer (GSI_ChemGuess_Bundle(it),'amassi',ges_amassi,istatus );ier=ier+istatus
+    call GSI_BundleGetPointer (GSI_ChemGuess_Bundle(it),'amassj',ges_amassj,istatus );ier=ier+istatus
+    call GSI_BundleGetPointer (GSI_ChemGuess_Bundle(it),'amassk',ges_amassk,istatus );ier=ier+istatus
+
+    if (ier/=0) then
+       print*,"get first guess amassi-k failed! Stop"
+       stop
+    end if
+
+! Get pointer to required control variable
+call gsi_bundlegetpointer (wbundle,'amassi',cv_amassi,istatus)
+call gsi_bundlegetpointer (wbundle,'amassj',cv_amassj,istatus)
+call gsi_bundlegetpointer (wbundle,'amassk',cv_amassk,istatus)
+
+! ad of partitioning amass into aerosol speccies 
+do ic=1,naerosols
+   ier = 0.0
+   call GSI_BundleGetPointer (GSI_ChemGuess_Bundle(it),trim(aerosols(ic)),ges_aero,istatus );ier=ier+istatus
+   call gsi_bundlegetpointer (rval,trim(aerosols(ic)),sv_rank3,istatus);ier=ier+istatus
+   if (ier/=0) cycle
+
+   print*,"amass2aero_ad: sum= ",ic,trim(aerosols(ic)),sum(sv_rank3),sum(cv_amassi),sum(cv_amassj),sum(cv_amassk)
```

Comment:
Isn't this going to print a lot of stuff to stdout (2 lines per processor per aerosol species)? Consider removing/commenting out.